index.d.ts

```diff
@@ -1,288 +1,522 @@
+    debug: boolean
+  }>
+
+  export interface UseTableHooks<D extends object> {
+    columnsBeforeHeaderGroups: ((
+      flatColumns: Column<D>[],
+      instance: TableInstance<D>
+    ) => Column<D>[])[]
+    columnsBeforeHeaderGroupsDeps: ((
+      deps: any[],
+      instance: TableInstance<D>
+    ) => any[])[]
+    useMain: ((instance: TableInstance<D>) => TableInstance<D>)[]
+    useRows: ((rows: Row<D>[], instance: TableInstance<D>) => Row<D>[])[]
+    prepareRow: ((row: Row<D>, instance: TableInstance<D>) => Row<D>)[]
+
+    // Prop Hooks
+    getTableProps: ((instance: TableInstance<D>) => object)[]
+    getRowProps: ((row: Row<D>, instance: TableInstance<D>) => object)[]
+    getHeaderGroupProps: ((
+      headerGroup: HeaderGroup<D>,
+      instance: TableInstance<D>
+    ) => object)[]
+    getHeaderProps: ((
+      column: Column<D>,
+      instance: TableInstance<D>
+    ) => object)[]
+    getCellProps: ((cell: Cell<D>, instance: TableInstance<D>) => object)[]
   }
 
-  export interface Row<D = {}> {
-    index: number
+  export type UseTableColumnOptions<D extends object> = Accessor<D> &
+    Partial<{
+      columns: Column<D>
+      show: boolean | ((instance: TableInstance<D>) => boolean)
+      Header: ComponentType<TableInstance<D>> | ReactNode
+      Cell: ComponentType<Cell<D>> | ReactNode
+    }>
+
+  export interface UseTableInstanceProps<D extends object> {
+    columns: Column<D>[]
+    flatColumns: Column<D>[]
+    headerGroups: HeaderGroup<D>[]
+    headers: HeaderGroup<D>[]
+    flatHeaders: HeaderGroup<D>[]
+    rows: Row<D>[]
+    getTableProps: (props?: object) => object
+    prepareRow: (row: Row<D>) => any
+    rowPaths: string[]
+    flatRows: Row<D>[]
+    setRowState: (rowPath: string, updated: Function | any) => void
+    setCellState: (
+      rowPath: string,
+      columnID: IdType<D>,
+      updater: Function | any
+    ) => void
+  }
+
+  export interface UseTableHeaderGroupProps<D extends object> {
+    headers: Column<D>[]
+    getHeaderGroupProps: (props?: object) => object
+  }
+
+  export interface UseTableColumnProps<D extends object> {
+    id: string
+    isVisible: boolean
+    render: (type: 'Header' | 'Filter', props?: object) => any // FIXME
+    getHeaderProps: (props?: object) => object
+  }
+
+  export interface UseTableRowProps<D extends object> {
     cells: Cell<D>[]
-    getRowProps: (userProps?: any) => any
+    values: Record<IdType<D>, D>
+    getRowProps: (props?: object) => object
+    index: number
     original: D
-    path: any[]
-    values: any[]
-    depth: number
+    path: IdType<D>[]
+    subRows: Row<D>[]
+    state: object
   }
 
-  export interface UseExpandedRow<D = {}> {
-    subRows?: D[]
-    groupByID?: string | number
-    toggleExpanded?: () => any
-    isExpanded?: boolean
-    isAggregated?: boolean
+  export interface UseTableCellProps<D extends object> {
+    column: Column<D>
+    row: Row<D>
+    value: unknown
```

Comment:
I think that value should be replaced with:

```ts
cell: { value: any }
```

Reply:
I also think that `UseTableCellProps` should extends TableInstance.

It's documented as
```txt
Receives the table instance and cell model as props
```

Reply:
Based on the docs, this should be `value`. https://github.com/tannerlinsley/react-table/blob/master/docs/api.md#cell-properties

And the code indicates that this is the value directly. https://github.com/tannerlinsley/react-table/blob/master/src/hooks/useTable.js#L328

This interface is used by `Cell`

Reply:
Oh, I think we're talking about 2 separate things. I'm talking about the cell properties, you're referring to the props that the Cell option on column gets.

Reply:
https://github.com/tannerlinsley/react-table/blob/master/src/hooks/useTable.js#L355

this calls the cell renderer passing {...instance, cell, column, row }.

Reply:
ah, right.